File: src/main/java/aurora/jwt/decoder/internal/SecurityContextConverter.java
package aurora.jwt.decoder.internal;

import static aurora.jwt.common.util.ValidationUtils.notNullOrEmpty;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import aurora.jwt.common.dto.Authorization;
import aurora.jwt.common.dto.Identity;
import aurora.jwt.common.dto.Preferences;
import aurora.jwt.common.dto.SecurityContext;
import aurora.jwt.common.util.Base36BitmaskEncoder;
import aurora.jwt.decoder.internal.payload.AuthorizationDto;
import aurora.jwt.decoder.internal.payload.IdentityDto;
import aurora.jwt.decoder.internal.payload.PreferencesDto;
import aurora.jwt.decoder.internal.payload.SecurityContextDto;


public class SecurityContextConverter {
    private final Base36BitmaskEncoder base36BitmaskEncoder;

    public SecurityContextConverter(Base36BitmaskEncoder base36BitmaskEncoder) {
        this.base36BitmaskEncoder = base36BitmaskEncoder;
    }

    public SecurityContext convert(SecurityContextDto payload) {
        return new SecurityContext(
                toIdentity(payload.getIdentity()),
                toPreferences(payload.getPreferences()),
                toAuthorization(payload.getAuthorization()));
    }

    private Identity toIdentity(IdentityDto identityDto) {
        String userId = notNullOrEmpty(identityDto.getUserId(), "'userId' can not be null");
        String organizationId = notNullOrEmpty(identityDto.getOrganizationId(), "'organizationId' can not be null");
        return new Identity(userId, organizationId);
    }

    private Preferences toPreferences(PreferencesDto preferencesDto) {
        return new Preferences(preferencesDto.getLocale(), preferencesDto.getTimezone(), preferencesDto.getFileEncoding());
    }

    private Authorization toAuthorization(AuthorizationDto authorizationDto) {
        List<Integer> decodedOrganizationWideAssets =
                this.base36BitmaskEncoder.decode(authorizationDto.getOrganization());

        // projectId -> Array[assets]
        Map<String, List<Integer>> decodedProjectAssets =
                getProjectIdToSecuredAssetsMap(authorizationDto.getProjects());

        return new Authorization(decodedOrganizationWideAssets, decodedProjectAssets);
    }

    @SuppressWarnings("checkstyle:CyclomaticComplexity")
    private Map<String, List<Integer>> getProjectIdToSecuredAssetsMap(Map<String, String> projects) {
        if (projects == null || projects.isEmpty()) {
            return Collections.emptyMap();
        }

        Map<String, List<Integer>> decodedProjectAssets = new LinkedHashMap<>();
        for (Map.Entry<String, String> projectIdToEncodedAssets : projects.entrySet()) {
            String projectId = projectIdToEncodedAssets.getKey();
            List<Integer> decodedSecuredAssets =
                    this.base36BitmaskEncoder.decode(projectIdToEncodedAssets.getValue());

            decodedProjectAssets.put(projectId, decodedSecuredAssets);
        }
        return decodedProjectAssets;
    }
}

File: src/main/java/aurora/jwt/decoder/internal/SecurityContextConverter.kt
package aurora.jwt.decoder.internal

import aurora.jwt.common.dto.Authorization
import aurora.jwt.common.dto.Identity
import aurora.jwt.common.dto.Preferences
import aurora.jwt.common.dto.SecurityContext
import aurora.jwt.common.util.Base36BitmaskEncoder
import aurora.jwt.common.util.ValidationUtils
import aurora.jwt.decoder.internal.payload.AuthorizationDto
import aurora.jwt.decoder.internal.payload.IdentityDto
import aurora.jwt.decoder.internal.payload.PreferencesDto
import aurora.jwt.decoder.internal.payload.SecurityContextDto

class SecurityContextConverter(private val base36BitmaskEncoder: Base36BitmaskEncoder) {
    fun convert(payload: SecurityContextDto): SecurityContext {
        return SecurityContext(
            toIdentity(payload.identity),
            toPreferences(payload.preferences),
            toAuthorization(payload.authorization)
        )
    }

    private fun toIdentity(identityDto: IdentityDto?): Identity {
        val userId = ValidationUtils.notNullOrEmpty(identityDto!!.userId, "'userId' can not be null")
        val organizationId =
            ValidationUtils.notNullOrEmpty(identityDto.organizationId, "'organizationId' can not be null")
        return Identity(userId, organizationId)
    }

    private fun toPreferences(preferencesDto: PreferencesDto?): Preferences {
        return Preferences(preferencesDto!!.locale, preferencesDto.timezone, preferencesDto.fileEncoding)
    }

    private fun toAuthorization(authorizationDto: AuthorizationDto?): Authorization {
        val decodedOrganizationWideAssets = base36BitmaskEncoder.decode(
            authorizationDto!!.organization
        )

        // projectId -> Array[assets]
        val decodedProjectAssets = getProjectIdToSecuredAssetsMap(
            authorizationDto.projects
        )
        return Authorization(decodedOrganizationWideAssets, decodedProjectAssets)
    }

    private fun getProjectIdToSecuredAssetsMap(projects: Map<String, String>?): Map<String, List<Int>> {
        if (projects == null || projects.isEmpty()) {
            return emptyMap()
        }
        val decodedProjectAssets: MutableMap<String, List<Int>> = LinkedHashMap()
        for ((projectId, value) in projects) {
            val decodedSecuredAssets = base36BitmaskEncoder.decode(value)
            decodedProjectAssets[projectId] = decodedSecuredAssets
        }
        return decodedProjectAssets
    }
}
